### PR TITLE
Maven Generator test harness

### PR DESCRIPTION
**Pull request by Karm**
See [Doc](https://github.com/Karm/quarkus-startstop/blob/generating-apps/README.md#artifactgeneratortest)

### CODE REVIEW

**Comment by Karm** [comment]
```
Error Message

Application FULL_MICROPROFILE in NATIVE mode took 46 ms to get the first OK request, which is over 45 ms threshold. ==> expected: <true> but was: <false>
```

I have increased it to 50 ms now.

**Comment by rsvoboda** [comment]
I tried your branch and it failed a lot for me

**Issues with rss measurements** 
I see `java.lang.IllegalArgumentException: rssKb must be a positive long, was: -1`
It's because `pa = new ProcessBuilder("ps", "--no-headers", "-p", Long.toString(pid), "-o", "rss");` doesn't work for me.
macOS based 'ps' doesn't know `--no-headers`, but there is workaround to apply by adding `=` to define empty column name after `rss` - for example `ps -p 33071 -o rss=`
Please adjust the PR

**Native fails**
I get message like: org.opentest4j.AssertionFailedError: NATIVE log should not contain `ERROR' lines that are not whitelisted. See testsuite/target/archived-logs/io.quarkus.ts.startstop.StartStopTest/fullMicroProfileNative/native-build.log ==> expected: <false> but was: <true>
The reason is that I have jdk8 based native-image and it's complaining about class file version 55, expects version up to 52.0.
README just say `mvn clean verify -Ptestsuite` and nothing about the prerequisites.

Please enhance the README to specify what's needed.
Upstream didn't jump exclusively on jdk11 based native-image, default is still jdk8 based native-image (http://quay.io/quarkus/ubi-quarkus-native-image:19.3.1-java8)

@mjurc /@Karm should we go back to Java 8 as a base or we should be looking at Java 11 based native image only ? I have no problem going fully Java 11+

**JVM mode tests only**
Can I have just JVM only tests executed ? That would be the scope for acceptance pipeline.

**Comment by rsvoboda** [comment]
Preliminary diff for ps command, trying now
```diff
-            pa = new ProcessBuilder("ps", "--no-headers", "-p", Long.toString(pid), "-o", "rss");
+            pa = new ProcessBuilder("ps", "-p", Long.toString(pid), "-o", "rss=");
```

File: testsuite/src/it/java/io/quarkus/ts/startstop/utils/MvnCmds.java [hunk 1]
```diff
@@ -29,12 +31,22 @@
...
             new String[]{"java", "-jar", "target/quarkus-runner.jar"}
     }),
-    //  DEV(new String[][]{
-    //          new String[]{"mvn", "clean", "compile", "quarkus:dev"}
-    //  }),
+    DEV(new String[][]{
+            new String[]{"mvn", "clean", "quarkus:dev"}
+    }),
     NATIVE(new String[][]{
             new String[]{"mvn", "clean", "compile", "quarkus:native-image", "-Pnative"},
```
Comment (rsvoboda) [inline testsuite/src/it/java/io/quarkus/ts/startstop/utils/MvnCmds.java]: `quarkus:native-image` is no longer suggested way - see https://github.com/quarkusio/quarkus/blob/master/devtools/maven/src/main/java/io/quarkus/maven/NativeImageMojo.java#L41
The same for probably for`quarkus:build`

This TS should be using ways described in https://quarkus.io/guides/building-native-image
I would go with `mvn clean package -Pnative -DskipTests -DskipITs`

Similar adjustment should happen to `new String[]{"mvn", "clean", "compile", "quarkus:build"},` line

Comment (rsvoboda) [reply testsuite/src/it/java/io/quarkus/ts/startstop/utils/MvnCmds.java]: Running `mvn clean verify` will probably hit some issues, maybe wotrh trying :-D
I filed few when playing with https://github.com/rsvoboda/rsvoboda-playground/blob/master/quarkus-extensions-combinations/foo.java

**Comment by rsvoboda** [comment]
I have apparently slooooow machine:

[ERROR]   StartStopTest.fullMicroProfileNative:173->testRuntime:134 Application FULL_MICROPROFILE in NATIVE mode took 713 ms to get the first OK request, which is over 50 ms threshold. ==> expected: <true> but was: <false>

[ERROR]   StartStopTest.jaxRsMinimalNative:159->testRuntime:134 Application JAX_RS_MINIMAL in NATIVE mode took 589 ms to get the first OK request, which is over 35 ms threshold. ==> expected: <true> but was: <false>

So the only way for me is to manually adjust `threshold.properties` or can i override it from command line ?

Maybe we could keep native images, I guess they are cleaned as part of https://github.com/quarkus-qe/quarkus-startstop/blob/master/testsuite/src/it/java/io/quarkus/ts/startstop/StartStopTest.java#L131

And sorry for mixing feedback for already merged stuff and this PR, I just didn't have time to try it earlier.

File: testsuite/src/it/java/io/quarkus/ts/startstop/ArtifactGeneratorTest.java [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+            // Config, see app-generated-skeleton/README.md
+            String appPropsSrc = BASE_DIR + File.separator + Apps.GENERATED_SKELETON.dir + File.separator + "application.properties";
+            String appPropsDst = appDir + File.separator + "src" + File.separator + "main" + File.separator + "resources" + File.separator + "application.properties";
+            Files.copy(Paths.get(appPropsSrc),
+                    Paths.get(appPropsDst), StandardCopyOption.REPLACE_EXISTING);
+
+            // Run
+            LOGGER.info("Running...");
+            runLogA = new File(logsDir + File.separator + (warmUp ? "warmup-dev-run.log" : "dev-run.log"));
+            pA = runCommand(MvnCmds.DEV.mvnCmds[0], appDir, runLogA);
```
Comment (rsvoboda) [inline testsuite/src/it/java/io/quarkus/ts/startstop/ArtifactGeneratorTest.java]: Why did you go with dev mode and not with building jar ?
I'm looking for reasoning.

**Comment by rsvoboda** [comment]
`Logs.logMeasurements` should also log to the standard output, not just to log file.
So it's easier to see the numbers and users do not need to jump between places.

When mentioning logging - jdk based logging on 2 lines is not great, I would prefer jboss-logging.

**Comment by rsvoboda** [comment]
Back to the slowness, when running `app-jax-rs-minimal/target/quarkus-runner` and `app-full-microprofile/target/quarkus-runner` I get under defined threshold.

My system is apparently recovering from huge load generated by native-image execution so some wait time before running `pA = runCommand(mvnCmds.mvnCmds[1], appDir, runLogA);` could help with stability of reported results.
Another option would be to perform several runs (e.g. 10), eliminate the maximum and the minimum, calculate average from the remaining `n-2` runs.

**Comment by Karm** [comment]
run tests

**Comment by Karm** [comment]
@rsvoboda @Ladicek  @mjurc  
Thank you for looking into it. Let me address your comments:

> **Issues with rss measurements**
> I see java.lang.IllegalArgumentException: rssKb must be a positive long, was: -1
> It's because pa = new ProcessBuilder("ps", "--no-headers", "-p", Long.toString(pid), "-o", "rss");
> doesn'twork for me.
> macOS based 'ps' doesn't know --no-headers

MacOS was simply not supported; README talked just Linux and Windows. Thanks for the Mac friendly ```ps``` suggestion; I added it. You ca test again on Mac.

> **Native fails**
> I get message like: org.opentest4j.AssertionFailedError: NATIVE log should not 
> contain ERROR' lines that are not whitelisted. See testsuite/target/archived-logs/
> io.quarkus.ts.startstop.StartStopTest/fullMicroProfileNative/native-build.log ==> expected:
> <false> but was: <true> The reason is that I have jdk8 based native-image and
> it's complaining about class file version 55, expects version up to 52.0.
> README just say mvn clean verify -Ptestsuite` and nothing about the prerequisites.

Java 11+ is required. The TS is using Java 9+ APIs etc.. Thank you for spotting it. I altered the README.md to clearly state the required JDK. The current implementation runs the test apps and the generated app with the same Java and the TS itself. If it is necessary to use JDK 8 at some point, we can enhance the TS to us multiple Javas.

> **JVM mode tests only**
> Can I have just JVM only tests executed ? That would be the scope for acceptance pipeline.

It is possible to do this now, use profile: ```-Ptestsuite-no-native```.

> **quarkus:native-image is no longer suggested way** - see 
> The same for probably for quarkus:build
> This TS should be using ways described in https://quarkus.io/guides/building-native-image
> I would go with mvn clean package -Pnative -DskipTests -DskipITs

This is somewhat tricky. I did not choose the phase arbitrarily, but because it was what worked for the big beefy app that packages additional resources (JWT keys). The app-jax-rs-minimal can use the simple profile, but the full-microprofile cannot without throwing ```java.lang.ClassNotFoundException: com.example.quarkus.client.Service``` during the build.

TL;DR: I altered the JVM mode, but I keep the native image maven command as it is until I establish what would work for both apps. Note that the apps run real code that does stuff, they are not empty skeletons.

Ad ```-DskipTests -DskipITs``` I have no idea what you mean - there are no tests in those apps, unit or integration. The apps are functional demos to be built and executed.

> **Running mvn clean verify will probably** hit some issues, maybe wotrh trying :-D
> I filed few when playing with ...

Are you suggesting this for the empty mvn generated skeleton or for those pre-configured demo apps? Let's describe it in a GitHub issue and follow up in the next PR.

> **I have apparently slooooow machine:...**
> So the only way for me is to manually adjust threshold.properties
> or can i override it from command line ?

Yes. See the README, it has been there since the initial PR:  See https://github.com/quarkus-qe/quarkus-startstop#thresholds  

> ```pA = runCommand(MvnCmds.DEV.mvnCmds[0], appDir, runLogA);```
> Why did you go with dev mode and not with building jar ?
> I'm looking for reasoning.

The idea was the user experience. When you run the generator, it gives you a one clear suggestion:

```
=============================================================================
Your new application has been created in /tmp/app-generated-skeleton
Navigate into this directory and launch your application with mvn quarkus:dev
Your application will be accessible on http://localhost:8080
=============================================================================

------------------------------------------------------------------------
BUILD SUCCESS
------------------------------------------------------------------------
```

So I run what it suggests. We can broaden it to other mvn phases if you find it pertinent. I suggest you open a GitHub issue.

> Logs.logMeasurements should also log to the standard output, not just to log file.
> So it's easier to see the numbers and users do not need to jump between places.
> When mentioning logging - jdk based logging on 2 lines is not great, 
> would prefer jboss-logging.

The last commit also spits the measurement into the log, e.g.

```
Mar 08, 2020 2:23:47 PM io.quarkus.ts.startstop.utils.Logs logMeasurements
INFO: 
App,Mode,buildTimeMs,timeToFirstOKRequestMs,startedInMs,stoppedInMs,RSSKb,FDs
FULL_MICROPROFILE,JVM,8763,1432,1124,20,178492,307
```

As for using of jboss-loggin: The current logging and logs archival facility is sufficient for the purpose of the initial release of the TS. I opened I GitHub issue for covering logging refactoring:
https://github.com/quarkus-qe/quarkus-startstop/issues/3

> My system is apparently recovering from huge load generated by native-image execution,
> some wait time before running pA = runCommand(mvnCmds.mvnCmds[1], appDir, runLogA);
> could help with stability of reported results.
> Another option would be to perform several runs (e.g. 10), eliminate the maximum and
> the minimum, calculate average from the remaining n-2 runs.

Ad "recovering": I can see nothing like that on Centos 8 / Fedora 25 / Windows 2019. I can resurrect my old Mac and we can pursue this on the GraalVM level if you wish. Although it is definitely in the scope of neither this TS nor this PR. The results are stable within some margin on baremetal hosts.

Ad ```several runs (e.g. 10)``` this is a statistical fallacy we have to avoid I believe. We would have to do histogram of runs and calculate SD and iteratively arrive at the number of runs that would be satisfactory (it could be hundreds depending on the stability of the system). We would have to treat outliers etc.

I suggest we keep it much simpler: Take the measurement into account only on selected stable baremetal boxes we control tightly.
 

I hope it helps and that we can move forward with this particular PR.
